[WIP] Implement CIP-0025

keyan-m opened the pull request:
As mentioned in [CIP-0068](https://github.com/cardano-foundation/CIPs/blob/5d32dc471d5c0f3a214e04ba415c674796732eb1/CIP-0068/README.md#metadata), the `metadata` field of the `reference_NFT` "is a low-level representation of the metadata, following closely the structure of CIP-0025." I'm still unsure whether it's a good idea to couple the two CIPs here. However, I think we should support CIP-0025 regardless.

keyan-m committed:
Implement `File` and some datatypes for MIME types

Currently, there is a separation between the image media type and other
ones. The reason for that was due to the requirement of an image MIME
type for the outer CIP-0025, while the one inside individual `File`s is
allowed to be any of the MIME types (a smarter datatype management can
probably address this).

Also, the implemented `Image` datatype might be too excessive. An
alternate solution could be utilizing the `elm/mimetype` package.

keyan-m committed:
Remove `assetName` and add `version`

keyan-m committed:
Implement `String.chunksOf` helper function

keyan-m committed:
Implement encoders/decoders for long strings

mpizenberg reviewed (commented):
Thanks a lot for starting this!
I've left a few comments, let me know what you think.

keyan-m commented on src/String/Extra.elm:
```diff
@@ -0,0 +1,25 @@
+chunksOf : Int -> String -> List String
+chunksOf chunkSize str =
+    let
+        go remainingStr acc =
```
Thank you for pointing this out, I wasn't aware of it. [This article](https://jfmengels.net/tail-call-optimization/) has gone into details of which locations are subject to TCO, and `let` bindings are not among them. I'll define a top level helper.

keyan-m commented on src/Cardano/Cip25.elm:
```diff
@@ -0,0 +1,815 @@
+to [IANA registry](https://iana.org/assignments/media-types/media-types.xhtml#image).
+
+Since the `image` field of CIP-0025 is required, and also must be one of the
+image types, this datatype leads to a more robust model (too excessive?).
```
I did mention that package in [one of my commit messages](https://github.com/mpizenberg/elm-cardano/pull/27/commits/d18f47583931e85777b2e4b831a17f5846899caa) (incorrectly 😅). One thing I don't like about e.g. `OtherImage` approach is that it leads to allowing two representations for "predefined" variants. However, this excessive list of constructors is also not pleasant, as I think most of them are never going to be needed and can only bloat the final bundles.

mpizenberg replied:
What do you think of making it much simpler by enabling potential multiple representations but providing a `canonicalize` function for that mime so that in practice we always call it, and the only case where it would not end in a canonical form would be when the user explicitly use the wrong constructor.

And even in that case, what are the potential downsides?
I suspect the encoder with end up with the same representation. So maybe the only downside is if someone tries to compare directly cip metadata on their own? Is that likely?

keyan-m commented on src/Cardano/Cip25.elm:
```diff
@@ -0,0 +1,815 @@
+The standard simply lays out a set of fields, some of which are optional.
+
+-}
+type alias Cip25 =
```
If I recall correctly, the reason I didn't use `Maybe` for `description` was similar to `files`, i.e. defaulting to an empty string/list in case of their absence (and omitting them from encoding if they're empty).

`mediaType` was an error on my part, I thought it was required. I'll make it a `Maybe` 👍

The standard points out that "A future version will introduce [schema.org](https://schema.org/)." While I frankly don't fully understand what they mean, looking through the website and also borrowing from [MeshJS](https://github.com/MeshJS/mesh/blob/e0b2a8cef3bb2095ec8afc87714eb07adfe59149/packages/module/src/common/types/AssetMetadata.ts#L11), it seems the idea is to support primary and secondary version numbers.

mpizenberg replied:
Sorry for having taken 7 months to answer ... :(

Regarding the version, let’s keep it to what the CDDL says, so just an `Int`.
As for the description and files. Empty list for files sounds fine, but empty string for description is sketchy. I’d prefer a Maybe for this one.

keyan-m committed:
Defined Recursion of `chunksOf` Top Level for TCO

keyan-m committed:
Reduce Supported Image MIME Types (For Now)

Potentially adding an "other" variant in future.

keyan-m committed:
Allow for Arbitrary Fields in `File`s and `Cip25`

I've modeled the arbitrary fields using `otherProps` accessors in both
`File` and `Cip25`, with types of `Dict String Metadatum`. This could be
somewhat restrictive as it only allows for `String` keys.

keyan-m committed:
Add JSON Encoder for `Cip25`

keyan-m committed:
Resolve Compile Errors

keyan-m committed:
Implement JSON Decoder for `Cip25`

keyan-m committed:
Add/Refine Comments

keyan-m committed:
Merge branch 'main' into feature/cip25

Also removed all JSON helpers, as they don't really seem necessary.

keyan-m committed:
refine data modeling

* Relax MIME type modeling
* Make `description` explicitly optional
* Use a sum type to represent versions

keyan-m committed:
refactor: relax image modeling

Breaking down URI's might be a better approach, as it'd make it more
convenient for developers to perform hygiene check on them before making
requests.

keyan-m committed:
feat: support multiple policies / asset names

keyan-m committed:
feat(cip25): basic smart constructors

keyan-m committed:
fix(cip25): proper string to chunks of 64 bytes

Previous implementation worked on characters rather than bytes.

keyan-m committed:
feat(cip25): string cbor encoder/decoder helpers

Plus some optimizations and a roundtrip test.

keyan-m committed:
feat(cip25): first draft for `File` cbor helpers

keyan-m committed:
chore(cip25): better arrangement of code

keyan-m committed:
refactor(cip25): make `Cip25` datatype opaque

Also removed the `empty` smart constructor, and refined the other
constructors. They all default to V2, and insertions to V1 will fail if
given asset name bytes is not valid UTF-8 text.

keyan-m committed:
feat(cip25): add accessors

keyan-m committed:
chore(cip25): better naming

keyan-m committed:
feat(cip25): asset metadata cbor en/decoder

keyan-m committed:
feat(cip25): 1st draft for `Cip25` en/decoder fns

keyan-m committed:
feat(cip25): more test coverage, extended version

Observing mainnet CIP-25 instances, some have used the `"1.0"` string as
the value for the `version` field. This is not CIP-25 compliant, but
explicitly supporting it allows unchanging roundtrips.

So `Version` now has 3 constructors for V1: implicit, string, and int.
While I did not find any instances for `"1"`, it seemed reasonable to
support that too.

Removed support for `"2.0"`. I have not found a single CIP-25 with
version 2 on mainnet yet.

Tests might be a bit of a slop, but they seem to give a good coverage.

keyan-m committed:
tests(cip25): add tests missed in prev commit

keyan-m committed:
fix(cip25): attempt fixing ci

keyan-m committed:
chore(cip25): correct formatting

keyan-m committed:
test(cip25): asset metadata decoding from full tx

keyan-m committed:
fix(cip25): datatype modeling artifact

`Image` was implemented as an opaque newtype in an experimental, simple
`Uri` modeling. Now that `Uri` is just a `String` alias, `Image` was not
adding any value.
